libeos-updater-util/flatpak.c

```diff
@@ -906,9 +925,39 @@ squash_ref_actions_ptr_array (GPtrArray *ref_actions)
   for (i = 0; i < ref_actions->len; ++i)
     {
       FlatpakRemoteRefAction *action = g_ptr_array_index (ref_actions, i);
-      g_hash_table_replace (hash_table,
-                            g_object_ref (action->ref),
-                            flatpak_remote_ref_action_ref (action));
+      FlatpakRemoteRefAction *existing_action_for_ref = NULL;
+
+      /* A little trickier than just blindly replacing, there are special
+       * rules regarding "update" since it only updates an existing installed
+       * flatpak, as opposed to installing it.
+       *
+       * (1) "install" and "uninstall" always take priority over "update"
+       *     since "install" means "install or update" and "uninstall"
+       *     means "unconditionally remove".
+       * (2) "update" takes priority over "uninstall" if "uninstall" were
+       *     present, since it will not re-install the app again, but it
+       *     will update it if the user had it installed.
+       * (3) "update" does not take priority over "install", since the latter
+       *     subsumes it anyway.
+       */
+      existing_action_for_ref = g_hash_table_lookup (hash_table, action->ref);
+
+      if (action->type == EUU_FLATPAK_REMOTE_REF_ACTION_INSTALL ||
+          action->type == EUU_FLATPAK_REMOTE_REF_ACTION_UNINSTALL ||
+          existing_action_for_ref == NULL ||
+          existing_action_for_ref->type == EUU_FLATPAK_REMOTE_REF_ACTION_UPDATE)
+        {
+          g_hash_table_replace (hash_table,
+                                g_object_ref (action->ref),
+                                flatpak_remote_ref_action_ref (action));
+        }
+      else if (action->type == EUU_FLATPAK_REMOTE_REF_ACTION_UPDATE)
+        {
+          if (existing_action_for_ref->type == EUU_FLATPAK_REMOTE_REF_ACTION_UNINSTALL)
```

Comment:
This block implements `[uninstall X, update X]` → `[update X]`, right? I don’t think that’s right. It should compress to `[uninstall X]`, because the update entry does nothing if the app is not installed. I think it has to be that way, because if that list is split in two and applied in two separate runs of the flatpak updater (for example, the first entry exists on day X, then an update happens and the second entry is added on day X+ 5), the outcome is `[uninstall X]`, not `[update X]`.

Reply:
Ah yes, that's a good point. I'll change it around.